server/routes/sitemap.js

```diff
@@ -0,0 +1,141 @@
+import AWS from 'aws-sdk';
+import { SitemapAndIndexStream, SitemapStream } from 'sitemap';
+import * as stream from 'stream';
+import { promisify } from 'util';
+import { createGzip } from 'zlib';
+
+import { isProd } from 'utils/environment';
+import { Page, Pub } from 'server/models';
+import app from 'server/server';
+import { handleErrors } from 'server/utils/errors';
+import { getInitialData } from 'server/utils/initData';
+import { hostIsValid } from 'server/utils/routes';
+import { communityUrl, pubUrl, pageUrl } from 'utils/canonicalUrls';
+
+const s3 = new AWS.S3({ params: { Bucket: 'sitemaps.pubpub.org' } });
+const pipeline = promisify(stream.pipeline);
+const keyPrefix = isProd() ? 'prod' : 'dev';
+
+const uploadFromStream = (key) => {
+	const pass = new stream.PassThrough();
+	const params = { Key: key, Body: pass, ACL: 'public-read' };
+
+	s3.upload(params, (err) => {
+		if (err) {
+			console.error(err);
+		}
+		pass.end();
+	});
+
+	return pass;
+};
+
+const getSitemapKey = (community, { filename, index = 'index' } = {}) =>
+	`${keyPrefix}/${community.id}/${filename || `sitemap-${index}.xml`}.gz`;
+
+const shouldGenerateSitemapIndex = async (community) => {
+	try {
+		const metadata = await s3.headObject({ Key: getSitemapKey(community) }).promise();
+		// Re-generate the sitemap index if sitemap has not been updated within
+		// the last 24 hours.
+		return Date.now() - metadata.LastModified >= 8.64e7;
+	} catch (err) {
+		return true;
+	}
+};
+
+const maybeGenerateSitemapIndex = async (community) => {
+	if (!(await shouldGenerateSitemapIndex(community))) {
+		return false;
+	}
+
+	const communityQuery = {
+		where: {
+			communityId: community.id,
+		},
+	};
+	const hostname = communityUrl(community);
+	const pubs = await Pub.findAll(communityQuery);
+	const pages = await Page.findAll(communityQuery);
+	// By default, SitemapAndIndexStream will partition sitemaps every 45,000 entries.
+	const sitemapAndIndexStream = new SitemapAndIndexStream({
+		getSitemapStream: (i) => {
+			const sitemapPath = `./sitemap-${i}.xml`;
+			const sitemapStream = new SitemapStream({
+				hostname: hostname,
+			});
+
+			sitemapStream
+				.pipe(createGzip())
+				.pipe(uploadFromStream(getSitemapKey(community, { index: i })));
+
+			return [new URL(sitemapPath, hostname).toString(), sitemapStream];
+		},
+	});
+	const urls = pubs
+		.map((pub) => ({
+			url: pubUrl(community, pub),
+			changefreq: 'weekly',
+			priority: 1,
+		}))
+		.concat(
+			pages.map((page) => ({
+				url: pageUrl(community, page),
+				changefreq: 'monthly',
+				priority: 0.8,
+			})),
+		);
+
+	await pipeline(
```

Comment:
Very cool, I have never really figured out pipelines but this looks super elegant!